### DISCUSSION TITLE
GitHub Private Repos - Cross Org Stopped Working

### DISCUSSION THREAD

**Discussion by ekristen** [renovatebot/renovate#15978] [Request Help]
Original discussion - https://github.com/renovatebot/renovate/discussions/11797

@rarkins this has stopped working.  

I've tried both with hostType set and not set.

```
    "hostRules": [
        {
            "matchHost": "https://api.github.com/repos/org2/",
            "encrypted": {
                "token": "wcFMA...."
            }
        }
    ],
```

This host rules is set in org1/repo and is trying to reference repos in org2. 

Log entries 
- 695642718
- 695642576

```
{
  "datasource": "github-releases",
  "packageName": "org2/some-repo-name1",
  "err": {
    "message": "Could not resolve to a Repository with the name 'org2/some-repo-name1.",
    "stack": "Error: Could not resolve to a Repository with the name 'org2/some-repo-name1'.\n    at CacheableGithubReleases.getItems (/home/ubuntu/renovateapp/node_modules/renovate/dist/modules/datasource/github-releases/cache/cache-base.js:126:31)\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (internal/process/task_queues.js:95:5)\n    at async GithubReleasesDatasource.getReleases (/home/ubuntu/renovateapp/node_modules/renovate/dist/modules/datasource/github-releases/index.js:143:26)\n    at async getRegistryReleases (/home/ubuntu/renovateapp/node_modules/renovate/dist/modules/datasource/index.js:57:17)\n    at async fetchReleases (/home/ubuntu/renovateapp/node_modules/renovate/dist/modules/datasource/index.js:231:23)\n    at async getPkgReleases (/home/ubuntu/renovateapp/node_modules/renovate/dist/modules/datasource/index.js:285:34)\n    at async lookupUpdates (/home/ubuntu/renovateapp/node_modules/renovate/dist/workers/repository/process/lookup/index.js:50:51)\n    at async fetchDepUpdates (/home/ubuntu/renovateapp/node_modules/renovate/dist/workers/repository/process/fetch.js:44:21)\n    at async /home/ubuntu/renovateapp/node_modules/p-map/index.js:57:22"
  }
}
```

**Comment by rarkins** [comment]
Likely because of graphql being used and no longer matching the same url prefixes. We'll need to think of a new solution

**Reply by ekristen** [reply]
Is the rest API no longer being used at all? Let me know, happy to test other solutions.

**Reply by rarkins** [reply]
We are using GraphQL for github tags and releases now, and were able to reduce average github API calls per repo/per hour in the hosted app by about 30%. The reason is that Graphql allows sorting by timestamp, which means we eliminate needing to do a full refresh and instead paginate by date until we hit a timestamp we already have (which is one page, in nearly every case).

I think we may need a solution where tokens can be specified in `packageRules` so that you can match them against package names or prefixes and pass them to the stack that way (rather than them being host-based).

**Reply by ekristen** [reply]
Understood. I like the idea of packageRules especially if I can regex the name or prefix it.

**Reply by rarkins** [reply]
BTW another problem even with REST was that GitHub's pagination points you to the repo ID after page 1, e.g. instead of `/repos/foo/bar/releases?page=2` it's something like `/repo/123185832/releases?page=2` which is very challenging to create hostRules for anyway.

**Reply by ekristen** [reply]
Yeah that was something we started to see a few months ago, I think I opened an issue on it. I like the packageRules approach. Let me know how I can help!

**Reply by ekristen** [reply]
Something like this would probably work really well and aligns better with the use of renovate and package rules anyways. I always felt the host rules felt out of place.

```
   "packageRules": [
        {
            "matchDatasources": [
                "gitub-releases"
            ],
            "matchPackagePatterns": [
                "^org1"
            ],
            "encrypted": {
                "token": "wcFMA..."
            }
        }
    ],
```

**Reply by rarkins** [reply]
Sure, in this case please open a feature request issue for supporting `token` in `packageRules`. We'll need to work out how it interacts with datasource/hostRules but should be ok to pass it as a datasource argument. PRs welcome too of course, if you have the time.

**Reply by ekristen** [reply]
Done https://github.com/renovatebot/renovate/issues/15997

**Comment by jmreicha** [comment]
I think I just ran into this. Is there a workaround or will we need to wait for the work in #15997 to be completed?
